### ISSUE TITLE
Leaflet map draggable marker event bug

### ISSUE THREAD

**Issue by shiro900** [Leaflet/Leaflet#4457]
Everytime I click on the map I create a draggable marker with a `drag` function. It works great, however everytime I release the mouse after I dragged some marker, the `click` event of the map is triggered. How can I stop that ?

```
map.on('click',
    function (e) {
        var pos = e.latlng;
        console.log('map click event');

        var marker = L.marker(
            pos,
            {
                draggable: true
            }
        );
        marker.on('drag', function (e) {
            console.log('marker drag event');
        });
        marker.on('click', L.DomEvent.stopPropagation);
        marker.addTo(map);
    }
);
```

**Edit:** `marker.on('click', L.DomEvent.stopPropagation);` This partially fixes the bug. Now if I drag the marker VERY quickly and release it, the bug still occurs, because the marker "follows" the mouse cursor in a speed slower than the speed with which I move my cursor, and so when the mouse is released, the mouse is not on the marker anymore, so the map `click` event is triggered.

**Edit 2** For the record I am using `Leaflet 1.0.0-rc1` and chrome `49.0.2623.112`

**Comment by balciseri** [comment]
What version are you using?
With the last version i can't reproduce it.
See: 
https://playground-leaflet.rhcloud.com/seda/edit

**Comment by shiro900** [comment]
I reproduced it. Try to click and drag a marker and release it as fast as you can.

**Comment by IvanSanchez** [comment]
I have a feeling this might be a duplicate of #4458, but let's leave both them open for now (and hope a fix for both happens)

**Comment by IvanSanchez** [comment]
I can reproduce with Chromium 50, but not with Firefox 45.

**Comment by shiro900** [comment]
If someone sees this and wants a temporary solution, take a look at the answer here:
http://gis.stackexchange.com/questions/190049/leaflet-map-draggable-marker-events

**Comment by balciseri** [comment]
I'm wondering why i can't reproduce the Click event beeing fired after i end the drag of the marker, i also use Chrome 50.0.2661.75

**Comment by shiro900** [comment]
You have to do the dragging and release so fast, that when your mouse is released the cursor is not on top of the marker.

**Comment by balciseri** [comment]
Ok i can reproduce it if i'm really fast, but reading your opening issue you said that this happens everytime this happens to you after a release of the mouse when you end a marker drag. That's why i was confused.

**Comment by shiro900** [comment]
Oh I see what you are saying. I actually HAVE to use `marker.on('click', L.DomEvent.stopPropagation);`, otherwise I get a map click event even if I click on the marker ! I am using `Leaflet 1.0.0-rc1` and chrome 49.0.2623.112`. My chrome is a couple of months outdated.

**Comment by balciseri** [comment]
A ok. As @IvanSanchez and @yohanboniface told me at https://github.com/Leaflet/Leaflet/issues/4458
This is not a bug but a feature. 
And you found the correct solution to this!

**Comment by balciseri** [comment]
> What version are you using?
With the last version i can't reproduce it.
See: 
https://playground-leaflet.rhcloud.com/seda/edit

Not sure if it's because i just upgraded crome from  50.0.2661.75 to 50.0.2661.87 but now the playground snippet is acting funny.

After the page loads if i click on the map nothing happens. I need first to drag the map then it starts to work ( aka is registering the click evend and inserting the marker when i click on the map). And i also started to get that bug that shiro900 was talking about.
Basically sometimes when i start dragging the marker it moves above my "hand cursos" so when i end the drag the cursor is on the map and not on the marker and a click event is firing.

Also after i dragged the marker again nothing happenes when i click, but on the second click it starts to work again. I'm not sure what's going on.


On firefox this is not happening.

**Comment by Shemeikka** [comment]
Hi,

I have a similar problem and it might be related to #4458 but since it's closed and this is open, I thought I would comment here.

When circle or cirleMarker has a popup binded, this popup will launch after dragging the marker. If I read right, then this should have been fixed, right? Maybe it has something to do with the way I'm doing the drag for circle and circleMarker.

I made a new playground that demonstrates this behavior:
[https://playground-leaflet.rhcloud.com/put/1/edit?html,js,output](https://playground-leaflet.rhcloud.com/put/1/edit?html,js,output)

Playground should be using Leaflet 1.0.0-rc.1+e2caf40 and I'm using Chrome 50.0.2661.102 m

**Comment by yohanboniface** [comment]
Maybe use [Leaflet.Path.Drag](https://github.com/w8r/Leaflet.Path.Drag) plugin for handling circle dragging?
Related #4628 and #4638.

**Comment by yohanboniface** [comment]
@balciseri should we close here?

**Comment by Shemeikka** [comment]
Yeah,

I tried that plugin but it didn't work at all.

**Comment by yohanboniface** [comment]
See #4628 and #4638 for inspiration if you wanna go by foot ;)

**Comment by Shemeikka** [comment]
The point is that the dragging works but the popup is fired after drag ends. I'm not sure if its because the way I'm doing the drag or is this a leaflet bug.

**Comment by yohanboniface** [comment]
> I'm not sure if its because the way I'm doing the drag or is this a leaflet bug. 

Neither. Your browser will issue a `click` event when you finish your drag and `mouseup`. And this `click` will trigger the popup if you binded one to it. Leaflet is smart enough to prevent this click when Leaflet itself is managing the drag, but if you do it by yourself it's out of its jurisdiction.
You can try to duck simulate, but this will need anyway a sort of handler, as in #4628.
One workaround if you don't want to go the North Face may be just to have you own way of binding the popup that takes care of that (i.e. when dragging set any flag anywhere, and on `click` check this flag before opening the popup).

**Comment by Shemeikka** [comment]
Okay,

Thanks for your quick replies. Javascript is not my strong point and I'm a little bit lost here. I need to do some reading about events and stuff or I could just try to get that drag-plugin to work.

**Comment by mtafel** [comment]
Here is my experience. Leaflet 1.0.3. 

I've a marker that is fixed at a certain length _n_ from a point, used for rotating another icon (_n_  essentially being a radius). 

Its quite easy to reproduce the unwanted map click event: mousedown on marker --> drag --> cursor/pointer/finger leaves the marker (due to _n_) and is now over the map --> mouseup (dragend) --> map fires click 

Is the behavior I describe the intended behavior?
 
No worries if it is, I've implemented the workaround that @shiro900 describes in his gis.stackoverflow post to get around the click.

**Comment by nrenner** [comment]
My use-case is drawing no-go circles with Leaflet.Editable while creating or editing a route with Leaflet.Routing. When dragging a circle edit handle (vertex marker), often a click event is fired on the map after drag, creating an unwanted route marker.

This happens when the vertex marker (a DivIcon) is dragged near the edge, which is more likely for such smaller DivIcons than the default marker icon.

Here is a demo with buttons for running reproducible automated tests for both near-edge and fast dragging scenarios (in Chrome, not Firefox):

http://plnkr.co/edit/awq40pnvEESdlIbfcHX5?p=preview

### near edge (offset)

Dragging a marker near the edge often leaves the cursor outside the marker after drag, thus the up and click events are fired on the map instead of the marker. You can also see the cursor wandering off the marker when dragging multiple times without moving the mouse in between.

The issue seems to be that `offset` is subtracted from `_startPos` before adding it later, thereby neutralizing the first move:
https://github.com/Leaflet/Leaflet/blob/eb9572800df7a6703423e31defebde04e638c6d7/src/dom/Draggable.js#L151

Simply removing the offset subtract seems to fix this issue (see demo patch (a)), but I don't understand what the purpose of the offset subtract is or why it [has been added](https://github.com/Leaflet/Leaflet/commit/2d2fc74110f2fe7fbc06d7f6e0a2e90528ce46c6#diff-d8f00e9d1726fd6483110dc44c93d9f8R84) in the first place. So probably I'm  missing something?

### fast dragging

Some findings regarding the fast dragging issue:

- the above offset issue also adds to this one
- the position of the last move is not updated because the `requestAnimFrame`  
  https://github.com/Leaflet/Leaflet/blob/eb9572800df7a6703423e31defebde04e638c6d7/src/dom/Draggable.js#L169  
  is cancelled in `onUp`/`finishDrag` (99dd46b538093a83c6402a829daa17b8cd362dad):  
  https://github.com/Leaflet/Leaflet/blob/eb9572800df7a6703423e31defebde04e638c6d7/src/dom/Draggable.js#L213-L214
- ignoring the click after dragging doesn't work, because `target` is the map, so `_draggableMoved` is checking the `_moved` flag of the map instead of the marker  
  https://github.com/Leaflet/Leaflet/blob/5ce473bab65be772e0cc6e25020c95ffba2fe5b2/src/map/Map.js#L1292-L1293

The root causes seem to be:

- last move, up and click events created together at the same time (according to Event.timeStamp) when dragging fast - see "timeStamp:" in console log when dragging fast manually in the demo - and up is dispatched before the next animation frame
- click target seems to be always the same as up target

Don't really know how to fix this: 
- not using `requestAnimFrame` and instead calling `_updatePosition` synchronously would fix this issue   (see demo first patch (b)), but that's probably not what we want?
- making sure the final position is set in any case (see demo second patch (b)) would fix the marker position but not avoid the click event on the map, as it is too late in the up handler, the click target is already set to the same as up

**Comment by johnd0e** [comment]
Ok, my explanation is different.
The root cause of the issue is non-optimal implementation of draggable: instead of using native drag-n-drop we simulate this with on `mousemove`.

Anyway, I suppose the issue can be fixed by setting element's correct final position in `finishDrag`.

**Comment by johnd0e** [comment]
> Simply removing the offset subtract seems to fix this issue (see demo patch (a)), but I don't understand what the purpose of the offset subtract is or why it [has been added](https://github.com/Leaflet/Leaflet/commit/2d2fc74110f2fe7fbc06d7f6e0a2e90528ce46c6#diff-d8f00e9d1726fd6483110dc44c93d9f8R84) in the first place. So probably I'm missing something?

You're missing the code _before_ was in `onDown`, and _after_ it moved to `onMove`.
Still, it seems to me that current code is wrong - it's impossible to get correct `_startPos` in `onMove`.

> * the position of the last move is not updated because the `requestAnimFrame`

Haven't checked that myself, but it's easy to fix anyway, adding `DomUtil.setPosition(this._element, this._newPos);` to `finishDrag`.

[These 2 fixes](https://github.com/johnd0e/Leaflet/commit/16298204199fb02e91c19e762e75ddac3be40109) improve the things, but..

> * making sure the final position is set in any case (see demo second patch (b)) would fix the marker position but not avoid the click event on the map, as it is too late in the up handler, the click target is already set to the same as up

May be it's not possible to fix click in this way at all.
Anyway, it's obvious that current  implementation of draggable in not optimal: we simulate dragging on `mousemove`, instead of using native drag-n-drop.

**Comment by johnd0e** [comment]
@shiro900 @Shemeikka @mtafel @nrenner 
Please check this solution: #7446

**Comment by nrenner** [comment]
> You're missing the code before was in _onDown, and after it moved to _onMove.

I'm aware of that, but subtracting the offset is and was also wrong in `_onMove`. Probably it was necessary at some point and it's a forgotten leftover of some refactoring during development. Not understanding why it was added just makes me a little uneasy.

> Still, it seems to me that current code is wrong - it's impossible to get correct _startPos in _onMove.

I also think it's wrong and that `_onDown` is the better place, but I can't follow you why it should not be possible. `_startPos` is only set on the first move when the element has [not moved](https://github.com/Leaflet/Leaflet/blob/436430db4203a350601e002c8de6a41fae15a4bf/src/dom/Draggable.js#L154) yet. So the element's position is still the same as in `_onDown`.

**Comment by johnd0e** [comment]
> So the element's position is still the same as in _onDown.

Indeed. You are right.
